keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1158 @@
+leaf to a trust anchor.  This might be referred to as a “root certificate” in
+other contexts, but nothing actually forces a trust anchor to be a **root**;
+your trust anchor might be an intermediate certificate.  Additionally, note that
+trust anchors are contextual; depending on the identity that a client or server
+presents to you, you might wish to select a different trust anchor set.
+
+This KEP introduces the certificates.k8s.io TrustAnchorSets cluster-scoped
+object, a vehicle for in-cluster certificate signers to communicate their trust
+anchors to workloads.
+
+Additionally, this KEP introduces a new trustAnchorSets kubelet projected volume
+source that gives workloads easy filesystem-based access to trust anchor sets
+that they might need.
+
+## Motivation
+
+In the absence of a standard mechanism for trust distribution, a few different
+approaches have emerged:
+
+* The trust anchors necessary for connecting to kube-apiserver are distributed
+  using a ConfigMap that is automatically injected into every namespace.
+
+* Kubernetes webhooks and aggregated API servers are configured with a static
+  trust anchor set embedded in their config (which complicates rotation).
+
+* * At least one signer implementation uses a [cluster-scoped
+  CRD](https://cloud.google.com/traffic-director/docs/security-proxyless-setup#:~:text=Save%20the%20following%20TrustConfig%20YAML%20configuration%20to%20tell%20your%20cluster%20how%20to%20trust%20the%20issued%20certificates%3A)
+  to distribute trust anchors to workloads.
+
+Each of these approaches has downsides:
+
+* Creating a ConfigMap in every namespace requires granting over-broad
+  permissions to the responsible controller (create ConfigMaps in **all**
+  namespaces).
+
+* Static trust anchor sets embedded directly in configuration complicate the job
+  of rotating the backing CA's roots of trust.
+
+* CRDs work well, but are inappropriate for use by core APIs.
+
+Each of these approaches would be simplified by an in-tree mechanism for
+distributing trust anchor sets.
+
+### Goals
+
+* Provide a Kubernetes data type that holds X.509 trust anchors for use wherever
+  a set of X.509 trust anchors is needed.
+
+* Make it easy for certificates/v1 signer implementations to publish their
+  relevant trust anchors using a TrustAnchorSets object.
+
+* Make it easy for workloads that need access to the trust anchors that back a
+  particular certificates/v1 signer to access them within the container
+  filesystem.
+
+* Allow cluster operators to configure webhook and aggregrated API server
+  backends using that use TrustAnchorSets instead of inline PEM data.
+
+### Non-Goals
+
+* Distribute a default set of "system" trust anchors using TrustAnchorSets.
+  (This is a natural extension, though).
+
+* Handle trust anchors expressed in other forms than PEM-wrapped, DER-formatted
+  X.509.
+
+## Proposal
+
+This proposal is centered around a new cluster-scoped TrustAnchorSets resource,
```

Comment:
If we call this new API `ClusterTrustAnchorSets` then we leave room for a future namespaced TrustAnchorSets API, perhaps a CRD implementation. Different namespaces might want their own trust anchors.

Reply:
That's a good idea --- I'll make that change.